v2/sourcedb-to-spanner/terraform/samples/sharded-bulk-migration/shardingConfig.json

```diff
@@ -0,0 +1,51 @@
+{
+  "configType": "dataflow",
+  "shardConfigurationBulk": {
+    "schemaSource": {
+      "dataShardId": "",
+      "host": "",
+      "user": "",
+      "password": "",
+      "port": "",
+      "dbName": ""
+    },
+    "dataShards": [
+      {
+        "dataShardId": "data-shard1",
+        "host": "10.1.1.1",
+        "user": "username",
+        "password": "password",
+        "port": "3306",
+        "dbName": "",
+        "databases": [
+          {
+            "dbName": "db1",
+            "databaseId": "logicaldb1",
+            "refDataShardId": "data-shard1"
+          },
+          {
+            "dbName": "db2",
+            "databaseId": "logicaldb2",
+            "refDataShardId": "data-shard1"
+          }
+        ]
+      },
+      {
+        "dataShardId": "data-shard2",
+        "host": "10.1.1.2",
+        "user": "username",
+        "password": "password",
+        "port": "3306",
+        "dbName": "",
+        "databases": [
+          {
+            "dbName": "db3",
+            "databaseId": "logicaldb3",
+            "refDataShardId": "data-shard2"
+          }
+        ]
+      }
+    ]
+  }
+}
```

Comment:
This should be native TF configuration and the JSON should be generated from that.

The biggest advantage is that in TF you define the objects declaratively so the user has all the information needed to construct an input from the TF definition itself.

https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/main/v2/datastream-to-spanner/terraform/samples/mysql-sharded-end-to-end/variables.tf#L76

Reply:
Our official documentation refers to launching it via gcloud which already uses this file. They will likely have this config already created. This was done to avoid recreation of this and avoid the extra step. Also a single config file that gcloud/terraform anything can use, and not generate new config based on tool.

Reply:
> They will likely have this config already created. This was done to avoid recreation of this and avoid the extra step.

I don't agree. I think you are confusing both of these things. Customer would likely use **one of** gCloud or Terraform to run the job - so there is no notion of an extra step or recreation here IMO. 

> Also a single config file that gcloud/terraform anything can use, and not generate new config based on tool.

This is **not** in fact a single global configuration. There are a lot of other configurations that are added in the gCloud command in addition to this config file. The **true** configuration file - which gives the complete snapshot of what a user configured, is the user's `tfvars` file.

Due to its many benefits around state management, configuration isolation and debuggability - we should in general encourage users towards using Terraform wherever possible, just like we used to encourage users to run via SMT instead of running via gCloud etc earlier. 

The gCloud way was strictly a way to get early users onto the product while we did the Terraform work. This is also the reason why I am insisting on having a simple, single-job template in place as well. 

Once the Terraform work is merged, you should also consider changing the official documentation/recommendation.